### ISSUE TITLE
Bug in glm::pow for quaternions

### ISSUE THREAD

**Issue by Fro116** [g-truc/glm#346]
In glm/gtx/quaternion.inl, the function pow (reproduced below) provides incorrect results. For example, glm::pow(glm::quat(0,1,0,0), 1.0f) returns the quaternion (1, nan, nan, nan). A fix is provided below

#### original function
template <typename T, precision P>
	GLM_FUNC_QUALIFIER tquat<T, P> pow
	(
		tquat<T, P> const & x,
		T const & y
	)
	{
		if(abs(x.w) > (static_cast<T>(1) - epsilon<T>()))
			return x;
		T Angle = acos(y);
		T NewAngle = Angle * y;
		T Div = sin(NewAngle) / sin(Angle);
		return tquat<T, P>(
			cos(NewAngle),
			x.x * Div,
			x.y * Div,
			x.z * Div);
	}

#### fixed function
template <typename T, glm::precision P>
	GLM_FUNC_QUALIFIER tquat<T, P> pow
	(
		tquat<T, P> const & x,
		T const & y
	)
	{
    		if(std::abs(x.w) > (static_cast<T>(1) - glm::epsilon<T>())) {
        	return x;
    	}
    	T Angle = acos(x.w);
    	T NewAngle = Angle * y;
    	T Div = sin(NewAngle) / sqrt(x.x*x.x + x.y*x.y + x.z*x.z);
    	return glm::tquat<T, P>(
                            cos(NewAngle),
                            x.x * Div,
                            x.y * Div,
                            x.z * Div);
	}

**Comment by Groovounet** [comment]
Hi,

What "sqrt(x.xx.x + x.yx.y + x.z * x.z);" is supposed to do?

This code is supposed to be equivalent to:
sqrt(x.x + x.x + x.z * x.z);

Thanks,
Christophe

**Comment by Fro116** [comment]
Sorry, I copy/pasted the wrong code. The fixed function should really be

template <typename T, precision P>
GLM_FUNC_QUALIFIER tquat<T, P> pow
(tquat<T, P> const & x, T const & y)
{

//Raising to the power of 0 should yield 1

//Needed to prevent a division by 0 error later on
if(y > -epsilon<T>() && y < epsilon<T>())

    return tquat<T, P>(1,0,0,0);

//To deal with non-unit quaternions

T magnitude = sqrt(x.x * x.x + x.y * x.y + x.z * x.z + x.w *x.w);

//Equivalent to raising a real number to a power

//Needed to prevent a division by 0 error later on
if(x.w / magnitude > static_cast<T>(1) - epsilon<T>() && x.w / magnitude < static_cast<T>(1) + epsilon<T>())

    return tquat<T, P>(pow(x.w, y),0,0,0);

T Angle = acos(x.w / magnitude);

T NewAngle = Angle * y;
T Div = sin(NewAngle) / sin(Angle);

T Mag = pow(magnitude, y-1);

return tquat(cos(NewAngle) * magnitude * Mag, x.x * Div * Mag, x.y * Div * Mag, x.z * Div * Mag);
}

Basically, there were two problems with the original implementation: the acos line had the wrong argument, and the original implementation only worked for unit quaternions.

**Comment by Fro116** [comment]
I forgot the abs() in the second if-statement. Change it to 

> if(abs(x.w / magnitude) > static_cast<T>(1) - epsilon<T>() && abs(x.w / magnitude) < static_cast<T>(1) + epsilon<T>())
>

**Comment by Groovounet** [comment]
This issue is fixed in master branch for GLM 0.9.7 release thanks to you patch.

Thanks for contributing,
Christophe
